async-iterator.md

```diff
@@ -0,0 +1,904 @@
+}
+```
+
+When either the caller or background execution need to move the method forward, execution will resume from the right label because the state variable was set to `N` and the dispatching logic will jump to that label.
+
+Because there is a cost to starting background execution, suspending and resuming, the lowering pattern is optimized: if the task was already completed, we avoid that overhead and instead just move ahead with the result.
+
+
+#### Locals
+
+Although I won't go into much details on this, it is worth mentioning that all the local variables in the method are converted to fields on a compiler-generated type, so they maintain their values across suspensions/resumptions.
+
+The compiler also generates fields for:
+- method parameters,
+- the `state` value,
+- persisting `awaiterTemp`,
+- the `current` value, (see section on "yield return" suspension)
+- the machinery for background execution and continuation.
+
+
+### Generic type parameters (optional)
+
+Consider the following async-iterator method:
+```C#
+IAsyncEnumerable<T> GetValuesAsync<T>(Task<T> slowValue)
+{
+    T value = await slowValue;
+    yield return value;
+    yield return default(T);
+}
+```
+
+Because the method has a type parameter, the generated type will have one as well (let's call it `T2`). As part of rewriting the body of the method, all references to `T` will be converted to references to `T2`:
+```C#
+class UnspeakableType<T2>
+{
+    private void MoveNext()
+    {
+        /* dispatch block */
+        /* lowered `T2 value = await slowValue;` */
+        /* lowered `yield return value;` */
+        /* lowered `yield return default(T2); */
+    }
+
+    /* various fields and public members omitted */
+}
+IAsyncEnumerable<T> GetValuesAsync<T>(Task<T> slowValue)
+{
+    return new UnspeakableType<T>();
+}
+```
+
+
+#### Spilling (optional)
+
+Lowering an `await expr` as described about works well when it is an expression statement (`await expr;`) or an assignment statement (`x = await expr;`).
+But when it appears inside another expression, such as an invocation `Method1(Method2(), await expr)`, this lowering is not sufficient.
+
+This invocation needs to evaluate `Method2()`, put the result on the stack, get a result from awaiting `expr`, put that result on the stack, then call `Method1` with both arguments on the stack.
+
+The problem is that the lowering for `await expr` involves a suspension (a `return` statement). But returning loses information saved in the current stack frame.
+
+The solution is to do some additional lowering (as an earlier pass) to turn `Method1(Method2(), await expr)` into:
+```C#
+var temp1 = Method2();
+var temp2 = await expr;
+Method1(temp1, temp2);
+```
+
+This relies on locals (instead of the stack) to store the two results.
+So when we rewrite async or async-iterator methods, we can assume that all awaits have been spilled.
```

Comment:
But temp2 in this example doesn't actually need to be lifted to the state machine, right?  It should be able to remain a local.

Reply:
In `M1(M2(), await expr);` the result of `M2()` needs to be saved and restored during the `await` suspension. ([example](https://sharplab.io/#v2:D4AQDABCCMDcCwAocVoFYHIMxQEwQGEIBvJCcqHEADigDYIBZAChDoB4BLAOwBcA+CAFMAHgAcATgEoSZCvLnzyjaM0a5mUgDRQAnMPHTMiiAF8TJkABYmqnrwgid9iAE8Zxc4nkv1mklAA7BBwZhBIpkA==))

But in `M1(await expr, M2())` there is no need for that. We just put the result of the `await` on the stack, the result of `M2()` on the stack, then invoke `M1`. ([example](https://sharplab.io/#v2:D4AQDABCCMDcCwAocVoFYHIMxQEwQGEIBvJCcqHEADigDYIBZAChDoB4BLAOwBcA+CAFMAHgAcATgEoSZCvLnzyjaKwCcw8RIA0TXMylTMiiAF8TJkABYmqnrwgjd9iAE8Zxc4nkvG+j1AA7BBwZhBIpkA==))